<?xml version="1.0"?>
<!--
Copyright (c) 2018, STEREOLABS.

All rights reserved.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->
<launch>
    <!-- Camera Model and Name -->
    <arg name="camera_name"           default="zed" /> <!-- The name you want -->
    <arg name="camera_model"          default="zed" /> <!-- 'zed' or 'zedm' or 'zed2' -->
    <arg name="node_name"             default="zed_node" />

    <!-- Load SVO file -->
    <arg name="svo_file"              default="" /><!-- <arg name="svo_file" default="path/to/svo/file.svo"> -->
    <!-- Remote stream -->
    <arg name="stream"                default="" /> <!-- <arg name="stream" default="<ip_address>:<port>"> -->

    <!-- Publish ZED urdf -->
    <arg name="publish_urdf"          default="true" />

    <arg name="camera_id"             default="0" />
    <arg name="gpu_id"                default="-1" />

    <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_pos_x"             default="0.0" />
    <arg name="cam_pos_y"             default="0.0" />
    <arg name="cam_pos_z"             default="0.0" />

    <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="cam_roll"              default="0.0" />
    <arg name="cam_pitch"             default="0.0" />
    <arg name="cam_yaw"               default="0.0" />


    <!-- resize depth and rgb-->
    <arg name="depth_resample_factor"         default="1.0"/>
    <arg name="img_resample_factor"           default="1.0"/>

    <!-- resize depth and rgb-->
    <arg name="base_frame"            default="base_link"/>
    <arg name="publish_map_odom"      default="false"/>  <!-- map to odom -->
    <arg name="publish_tracking"      default="true"/>

    <arg name="point_cloud_freq"      default="15"/>


    <!-- ZED2 -->
    <arg if="$(eval camera_model=='zed2')" name="object_detection_enabled" default="false"/>

    <!-- ROS URDF description of the ZED -->
    <group if="$(arg publish_urdf)">
        <param name="zed_description"
               command="$(find xacro)/xacro '$(find zedbot_description)/urdf/zed_descr.urdf.xacro'
                        camera_name:=$(arg camera_name)
                        camera_model:=$(arg camera_model)
                        cam_pos_x:=$(arg cam_pos_x)
                        cam_pos_y:=$(arg cam_pos_y)
                        cam_pos_z:=$(arg cam_pos_z)
                        cam_roll:=$(arg cam_roll)
                        cam_pitch:=$(arg cam_pitch)
                        cam_yaw:=$(arg cam_yaw)"
        />
        <!-- uses zedm or zed2 imu to properly project zed data in robot frame (take zed orientation into account)--> 
        <!-- If  camera_model == zed (zed model does not have imu) pitch is supposed null-->
        <node name="zed_orientation_publisher" pkg="zedbot_bringup" type="zed_orientation_publisher" />
        
        <node name="zed_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen" required="true">
            <remap from="robot_description" to="zed_description" />
            <remap from="joint_states" to="/zed/zed_orientation_joint_states" />
        </node>
    </group>

    <node name="$(arg node_name)" pkg="zed_wrapper" type="zed_wrapper_node" output="screen" required="true"><!-- launch-prefix="valgrind" -->
        <rosparam file="$(find zed_wrapper)/params/common.yaml" command="load" />
        <rosparam file="$(find zed_wrapper)/params/$(arg camera_model).yaml" command="load" />

        <!-- Camera name -->
        <param name="general/camera_name"               value="$(arg camera_name)" />

        <!-- SVO file path -->
        <param name="svo_file"          value="$(arg svo_file)" />

        <!-- Remote stream -->
        <param name="stream"            value="$(arg stream)" />

        <!-- Camera ID -->
        <param name="general/zed_id"             value="$(arg camera_id)" />

        <!-- GPU ID -->
        <param name="general/gpu_id"             value="$(arg gpu_id)" />

        <!-- gle settings-->
        <param name="point_cloud_freq"          value="$(arg point_cloud_freq)" />
        <param name="pub_frame_rate"          value="15" />
        <param name="general/base_frame"          value="$(arg base_frame)" />
        <param name="depth/min_depth"             value="0.5" />
        <param name="depth/depth_downsample_factor" value="$(arg depth_resample_factor)" />
        <param name="video/img_downsample_factor"   value="$(arg img_resample_factor)" />
        
        <param name="pos_tracking/publish_tf"       value="$(arg publish_tracking)" />
        <param name="pos_tracking/publish_map_tf"   value="$(arg publish_map_odom)" />
	    <param name="pos_tracking/two_d_mode"       value="true" />
        <param name="pos_tracking/fixed_z_value"    value="0.0" /> <!-- z coord base_link -->

        <param if="$(eval camera_model=='zed2')" name="object_detection/od_enabled"  value="$(arg object_detection_enabled)" />

    </node>
</launch>
